### ISSUE TITLE
repo key message in charm-build confusing

### ISSUE THREAD

**Issue by hatched** [juju/charm-tools#394]
`build: Please add a `repo` key to your layer.yaml, with a url from which your layer can be cloned.`

layer.yaml:
```
includes: ['layer:basic', 'layer:metrics']  # if you use any interfaces, add them here
```

I'm not building a layer, I'm building a charm, why am I seeing this message?

**Comment by marcoceppi** [comment]
You're building a layer, which is then compiled into a charm. The `repo` key in the YAML identifies where the source for this layer (and charm) are stored.

**Comment by hatched** [comment]
I'm not making a layer though, I'm building a charm with custom metadata, hooks, assets, etc.

**Comment by marcoceppi** [comment]
Then don't use charm build.

**Comment by hatched** [comment]
I'm not sure I understand, to use layers I must use `charm build`, yes? But it is in it's "final form" so it can't be included in another charm... So is every charm that uses layers a layer itself? Or is it not a charm until it's been built if it uses layers, even if it includes hooks and the like?

**Comment by marcoceppi** [comment]
If you're using layers, than you're building a charm layer - doesn't matter if it has a hook directory, it's only a fragment of a charm. In which case you need to use `charm build` and you are writing a layer. A charm is either written from scratch or compile from a series of layers. If you look, for example, at the base layer, you'll see it has hooks and other file/folder fragements. Layers are not limited to just the reactive directory - in fact reactive framework and charm layers are not mutually exclusive.

In short, this is a layer despite what you may feel, and you should include a `repo` key to point to where the source for this layer is :wink:

**Comment by johnsca** [comment]
There's some confusion over the terminology here, which is [explained in more detail](https://jujucharms.com/docs/stable/developer-layers#what-are-layers?) in the docs. I think the source of the confusion is that, while in a technical sense "layer" is the correct term since the charm layer is technically a type of layer, and there are even some charm layers that can themselves be used as base layers as well, if you're running `charm build` on it, it really is specifically a "charm layer" in that usage.

@hatched: The reason why it's important to use the term "layer" here is to distinguish it from from the output of the `charm build` step, which is the "built charm", or just "the charm" in the usual usage of the term (i.e., the thing that's actually deployed).  The `repo` key is specifically there to act as a pointer back to the source layer for the charm, for the `charm pull-source` command, so that the charm can be built from scratch again (or at least just so those interested in seeing the un-built source can find it).

That said, I don't think there's anything lost by being more precise in the language here by saying "charm layer" rather than just "layer" (or possibly even "charm layer source").  Do you think that would be sufficiently less confusing, @hatched?  @marcoceppi?
